merkle-tree/hash-set/src/lib.rs

```diff
@@ -431,60 +432,102 @@ where
         for i in 0..self.capacity_values {
             let probe_index = (value.clone() + i.to_biguint().unwrap() * i.to_biguint().unwrap())
                 % self.capacity_values.to_biguint().unwrap();
             let probe_index = probe_index.to_usize().unwrap();
-            let index_bucket = unsafe { &mut *self.indices.as_ptr().add(probe_index) };
+            let index_bucket = unsafe { &*self.indices.as_ptr().add(probe_index) };
 
             match index_bucket {
-                // The visited hash set cell points to a value in the array.
                 Some(value_index) => {
-                    let value_index =
-                        usize::try_from(*value_index).map_err(|_| HashSetError::UsizeConv)?;
-                    if self.insert_into_occupied_cell(
-                        value_index,
-                        value,
+                    let value_bucket = self.by_value_index(
+                        usize::try_from(*value_index).map_err(|_| HashSetError::UsizeConv)?,
                         current_sequence_number,
-                    )? {
-                        return Ok(());
+                    );
+
+                    if let Some(value_bucket) = value_bucket {
+                        if &value_bucket.value_biguint() == value {
+                            return Ok(Some((value_bucket, *value_index)));
+                        }
                     }
                 }
                 None => {
-                    let value_bucket =
-                        unsafe { &mut *self.values.as_ptr().add(*self.next_value_index) };
-                    // SAFETY: `next_value_index` is always initialized.
-                    *index_bucket = Some(
-                        I::try_from(unsafe { *self.next_value_index })
-                            .map_err(|_| HashSetError::IntegerOverflow)?,
-                    );
-                    *value_bucket = Some(HashSetCell {
-                        value: bigint_to_be_bytes_array(value)?,
-                        sequence_number: None,
-                    });
-                    // SAFETY: `next_value_index` is always initialized.
-                    unsafe {
-                        *self.next_value_index =
-                            if *self.next_value_index < self.capacity_values - 1 {
-                                *self.next_value_index + 1
-                            } else {
-                                0
-                            };
-                    }
-                    return Ok(());
+                    return Ok(None);
                 }
             }
         }
 
-        Err(HashSetError::Full)
+        Ok(None)
     }
 
-    pub fn find_element(
+    /// iterate over a range of elements in the hash set
+    /// return the position of the first free value
+    /// We always have to iterate over the whole range to make sure that the value has not been inserted before
+    pub fn find_element_iter(
```

Comment:
Is there any reason behind creating a new method, instead of applying the changes to `find_element`?

Correct me if I'm wrong, but I tried to understand the difference between `find_element` and `find_element_iter`. My understanding so far is that `find_element_iter`:

* Returns a boolean indicating whether any value was inserted into that cell before. Good change and very good catch.
  * But I'm wondering - is there any reason for not including it in `find_element_iter`? Callers which don't care about that information could just skip it. Or we could provide a variant not returning that info (but still calling internally the method which returns it).
* `find_element` has a loop from 0 to capacity. `find_element_iter` has a loop bounded by a custom value, which is then randomly set to 20 in the other place. Why? My reasoning behind using capacity as a bound was to make sure that we don't miss any cell. I couldn't think of any other bound which could guarantee that. Do you have any in mind? And why is it 20?

Reply:
In general I don't see a reason why we can't unify these functions.
Just the return type is a bit different, so I implemented it this way for speed.

Reply:
I think because of the different return type it doesn't make sense to unify them.